contracts/StablecoinConverter.sol

```diff
@@ -108,4 +109,65 @@ contract StablecoinConverter is EpochTokenLocker {
     function tokenIdToAddressMap(uint16 id) public view returns (address) {
         return IdToAddressBiMap.getAddressAt(registeredTokens, id);
     }
+
+    struct TradeData {
+        address owner;
+        uint volume;
+        uint16 orderId;
+    }
+
+    function submitSolution(
+        uint32 batchIndex,
+        address[] memory owner,  //tradeData is submitted as arrays
+        uint16[] memory orderId,
+        uint128[] memory volume,
+        uint128[] memory prices,
+        uint16[] memory tokenIdForPrice
+    ) public {
+        require(
+            batchIndex == getCurrentStateIndex() - 1,
+            "Solutions are no longer accepted for this batch"
+        );
+        // checkTradersUtility();
+        // checkPriceConformity(prices);
+        // checkNorm(prices);
+        uint len = owner.length;
+        for (uint i = 0; i < len; i++) {
+            Order memory order = orders[owner[i]][orderId[i]];
+            require(order.validFrom <= batchIndex, "Order is not yet valid");
+            require(order.validUntil >= batchIndex, "Order is no longer valid");
+            // Assume for now that we always have sellOrders
+            uint128 sellVolume = volume[i];
+            uint128 buyVolume = uint128(
+                volume[i].mul(prices[findPriceIndex(order.buyToken, tokenIdForPrice)]) /
+                prices[findPriceIndex(order.sellToken, tokenIdForPrice)]
+            );
+            require(
+                sellVolume.mul(order.buyAmount) >= buyVolume.mul(order.sellAmount),
+                "limit price not met"
+            );
+            require(order.sellAmount >= sellVolume, "sellVolume bigger than specified in order");
+            orders[owner[i]][orderId[i]].buyAmount = uint128(
+                order.sellAmount.sub(sellVolume)
+                .mul(order.buyAmount) / order.sellAmount
+            );
+            orders[owner[i]][orderId[i]].sellAmount = uint128(order.sellAmount.sub(sellVolume));
+            addBalance(owner[i], tokenIdToAddressMap(order.buyToken), buyVolume);
+        }
+        //doing the substracts after doing all additions, in order to avoid negative values
+        for (uint i = 0; i < len; i++) {
+            Order memory order = orders[owner[i]][orderId[i]];
+            substractBalance(owner[i], tokenIdToAddressMap(order.sellToken), volume[i]);
+        }
+    }
+
+    function findPriceIndex(uint16 index, uint16[] memory tokenIdForPrice) public pure returns (uint) {
```

Comment:
This is cute, but somewhat inefficient...

Observe that you are fetching the price index for each order in the loop above (`submitSolution`). This means you have `O(numTokens * numOrders)` complexity while you could cache this token index info once and continue reusing it! This would reduce your `submitSolution` to `O(numOrders)`

Reply:
I agree that it is not highly inefficient. 

But I would like to continue like that for now and make this optimization later. The reason is that later on, once the function submitSolution is feature complete, I expect to have trouble with the stack deepness. I can only have 16 variables and I am already at 10.  Checking which values should be stored and which one should be recomputed should be done, once we have everything well defined.

Reply:
Ok then. Maybe we could include a comment or todo so this doesn’t get buried.